synapse/storage/databases/main/event_federation.py

```diff
@@ -1309,27 +1309,15 @@ async def record_event_failed_backfill_attempt(
             room_id: The room where the event failed to backfill from
             event_id: The event that failed to be backfilled
         """
-        if (
-            self.database_engine.can_native_upsert
-            and "event_failed_backfill_attempts"
-            not in self.db_pool._unsafe_to_upsert_tables
-        ):
-            await self.db_pool.runInteraction(
-                "record_event_failed_backfill_attempt",
-                self._record_event_failed_backfill_attempt_upsert_native_txn,
-                room_id,
-                event_id,
-                db_autocommit=True,  # Safe as its a single upsert
-            )
-        else:
-            await self.db_pool.runInteraction(
-                "record_event_failed_backfill_attempt",
-                self._record_event_failed_backfill_attempt_upsert_emulated_txn,
-                room_id,
-                event_id,
-            )
+        await self.db_pool.runInteraction(
+            "record_event_failed_backfill_attempt",
+            self._record_event_failed_backfill_attempt_upsert_txn,
+            room_id,
+            event_id,
+            db_autocommit=True,  # Safe as it's a single upsert
+        )
```

Comment:
Sanity check: is `event_failed_backfill_attempts` safe to upsert?

I'm not 100% on top of the logic behind `unsafe_to_upsert_tables`, but my understanding is: because (room_id, event_id) is a primary key for this table, there is implicitly a UNIQUE index on that pair of columns, making this upsert safe.

Reply:
Ahh, discussed here: [`63bec99` (#13589)](https://github.com/matrix-org/synapse/pull/13589/commits/63bec99dff683e68ae38837ef4bee837a90e1b22#r968246547)

Reply:
It [previously had an `unsafe_to_upsert_tables` check](https://github.com/matrix-org/synapse/pull/13589#discussion_r960436114) but was [removed](https://github.com/matrix-org/synapse/pull/13589#discussion_r961102072) after positing that it wasn't necessary since `unsafe_to_upsert_tables` is a static list of things we didn't create correctly in the first place.

We have `PRIMARY KEY (room_id, event_id)` for `event_failed_pull_attempts` which has implicit uniqueness 👍

> A primary key constraint indicates that a column, or group of columns, can be used as a unique identifier for rows in the table. 
> 
> *-- https://www.postgresql.org/docs/current/ddl-constraints.html#DDL-CONSTRAINTS-PRIMARY-KEYS*